### PR TITLE
[Merged by Bors] - feat: subgroups of Z/nZ-modules are submodules

### PR DESCRIPTION
**Pull request by llllvvuu**
Together with #9017 and #8965 we can convert group theory problems to linear algebra problems as in the
[homomorphism form of PFR](https://teorth.github.io/pfr/blueprint/sect0008.html).

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ZMod/Module.lean [hunk 1]
```diff
@@ -10,35 +10,49 @@ import Mathlib.Algebra.Module.LinearMap
 # The `ZMod n`-module structure on Abelian groups whose elements have order dividing `n`
 -/
 
-variable {n : ℕ} {M M₁ : Type*} [AddCommGroup M] [AddCommGroup M₁]
-  [Module (ZMod n) M] [Module (ZMod n) M₁]
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Module.lean:14]: We usually use `M` for modules, even if they're additive groups. Can you change this back to the old names?

(That also has the advantage of making the diff much smaller)

Comment (llllvvuu) [reply Mathlib/Data/ZMod/Module.lean:14]: Thanks, nice catch, fixed! (I'd accidentally thought "M" stood for "Monoid" - in my own code 😅)

File: Mathlib/Data/ZMod/Module.lean [hunk 2]
```diff
@@ -42,3 +42,36 @@ theorem toZModLinearMap_injective: Function.Injective <| toZModLinearMap n (M :=
...
+
+See also: `AddSubgroup.toIntSubmodule`, `AddSubmonoid.toNatSubmodule`. -/
+def toZModSubmodule : AddSubgroup M ≃o Submodule (ZMod n) M where
+  toFun S := { S with smul_mem' := fun c _ h ↦ ZMod.smul_mem (K := S) h c }
+  invFun := Submodule.toAddSubgroup
+  left_inv _ := rfl
+  right_inv _ := rfl
+  map_rel_iff' := Iff.rfl
+
+@[simp]
```
Comment (llllvvuu) [inline Mathlib/Data/ZMod/Module.lean:58]: I had the idea to copy these simp lemmas from `toIntSubmodule`, but is there a way to automate these?

File: Mathlib/Data/ZMod/Module.lean [hunk 3]
```diff
@@ -10,25 +10,25 @@ import Mathlib.Algebra.Module.LinearMap
 # The `ZMod n`-module structure on Abelian groups whose elements have order dividing `n`
 -/
 
-variable {n : ℕ} {M M₁ : Type*} [AddCommGroup M] [AddCommGroup M₁]
-  [Module (ZMod n) M] [Module (ZMod n) M₁]
+variable {n : ℕ} {M M₁ F S : Type*} [AddCommGroup M] [AddCommGroup M₁] [AddMonoidHomClass F M M₁]
+  [Module (ZMod n) M] [Module (ZMod n) M₁] { x : M } [SetLike S M] [AddSubgroupClass S M] {K : S}
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Module.lean:14]:
```suggestion
  [Module (ZMod n) M] [Module (ZMod n) M₁] [SetLike S M] [AddSubgroupClass S M] {x : M} {K : S}
```

**Review by eric-wieser — APPROVED** [review]
bors merge

Thanks!

**Pull request closed without merging**